File: src/main/java/domain/model/rlsol/Rlsol.java
package domain.model.rlsol;

import java.awt.geom.Point2D;
import java.util.List;

public class Rlsol {

    private final Point2D.Double[] points;

    /**
     * @param pu barometric pressure
     * @param tans intake air temperature
     * @param kfurl ~0.1037
     * @param pressure pressures in mbar
     */
    public Rlsol(double pu, double tans, double kfurl, List<Double> pressure) {
        points = new Point2D.Double[pressure.size()];
        double ps = pressure.get(0);

        for(int i = 0; i < points.length; i++) {
            double x = pressure.get(i);
            double y = rlsol(pu, ps, tans, 96, kfurl, x);
            ps = x;

            points[i] = new Point2D.Double(x,y);
        }
    }

    public Point2D.Double[] getPoints() {
        return points;
    }

    /**
     *
     * @param pu barometric pressure
     * @param ps intake manifold absolute pressure
     * @param tans intake air temperature
     * @param tmot coolant temperature
     * @param kfurl ~0.1037
     * @param plsol requested boost pressure
     * @return
     */
     public static double rlsol(double pu, double ps, double tans, double tmot, double kfurl, double plsol) {
        final double KFPRG = 70.0;
        double FPBRKDS = 1.016; // VE
        final double KFFWTBR = 0.02;
        final double VPSSPLS = 1.016; // pressure drop over the throttle plate

        double fho = pu/1013.0;

        double pirg = fho*KFPRG;

        double pbr = ps * FPBRKDS;
        double psagr = 250; // ?

        double evtmod = tans + (tmot - tans) * KFFWTBR;
        double fwft = (tans+673.425)/731.334;
        double ftbr = 273.0/(evtmod+273)*fwft;

        double fupsrl = kfurl * ftbr;

        double rfagr = Math.max(pbr-pirg, 0)*fupsrl*psagr/ps;

        return (plsol * fupsrl * FPBRKDS * VPSSPLS) - rfagr;
    }
}

File: src/main/kotlin/domain/model/rlsol/Rlsol.kt
package domain.model.rlsol

import domain.math.Point

class Rlsol(pu: Double, tans: Double, kfurl: Double, pressure: List<Double>) {
    val points: Array<Point>

    init {
        points = Array(pressure.size) { Point(0.0, 0.0) }
        var ps = pressure[0]
        for (i in points.indices) {
            val x = pressure[i]
            val y = rlsol(pu, ps, tans, 96.0, kfurl, x)
            ps = x
            points[i] = Point(x, y)
        }
    }

    companion object {
        fun rlsol(pu: Double, ps: Double, tans: Double, tmot: Double, kfurl: Double, plsol: Double): Double {
            val KFPRG = 70.0
            val FPBRKDS = 1.016
            val KFFWTBR = 0.02
            val VPSSPLS = 1.016

            val fho = pu / 1013.0
            val pirg = fho * KFPRG
            val pbr = ps * FPBRKDS
            val psagr = 250.0

            val evtmod = tans + (tmot - tans) * KFFWTBR
            val fwft = (tans + 673.425) / 731.334
            val ftbr = 273.0 / (evtmod + 273) * fwft

            val fupsrl = kfurl * ftbr
            val rfagr = maxOf(pbr - pirg, 0.0) * fupsrl * psagr / ps

            return (plsol * fupsrl * FPBRKDS * VPSSPLS) - rfagr
        }
    }
}
